src/control/modules/motion-control/PidMotionController.hpp

```diff
@@ -56,14 +102,85 @@ class PidMotionController {
-        // convert encoder ticks to rad/s
+        // update control targets
+        // in the future, we can get the rotation angle soccer wants and
+        // directly command that
+        // as our target, or integrate the rotational velocities given to us to
+        // create the target.
+        // For now though, we only do an angle hold when soccer is not
+        // commanding rotational velocities
+        // (this should help with strafing quickly)
+        // target_rotation += _targetVel[2] * dt;
+
+        // get sensor data
+        imu.getMotion6(&ax, &ay, &az, &gx, &gy, &gz);
+
+        // convert sensor readings to mathematically valid values
         Eigen::Vector4d wheelVels;
         wheelVels << encoderDeltas[0], encoderDeltas[1], encoderDeltas[2],
             encoderDeltas[3];
         wheelVels *= 2.0 * M_PI / ENC_TICKS_PER_TURN / dt;
 
+        auto bot_vel = RobotModel::get().WheelToBot * wheelVels;
+
+        // we have two redundent sensor measurements for rotation
+        // 32.8 comes from data sheet, units are LSB / (deg/s)
+        float ang_vel_gyro = (gz / 32.8f) * M_PI / 180.0f;
+        float ang_vel_enc = bot_vel[2];
+
+        // printf("%f\r\n", ang_vel_enc);
+        // std::printf("%f %f\r\n", ang_vel_gyro, ang_vel_enc);
+
+        // perform sensor fusion
+        // the higher this is, the more gyro measurements are used instead of
+        // encoders
+        float sensor_fuse_ratio = 1;
+        float ang_vel_update = ang_vel_gyro * sensor_fuse_ratio +
+                               ang_vel_enc * (1 - sensor_fuse_ratio);
+
+        // perform state update based on fused value, passed through a low
+        // passed filter
+
+        // so far noise on the gyro seems pretty minimal, that's why this filter
+        // is off
+        float alpha = 1.0;  // 0->1 (higher means less filtering)
+        angular_vel = (alpha * ang_vel_update + (1 - alpha) * angular_vel);
+        // current rotation estimate
+        rotation += angular_vel * dt;
+
+        // printf("%f\r\n", rotation * 180.0f / M_PI);
+
+        // velocity we are actually basing control off of, not the latest
+        // command
+        auto target_vel_act = _targetVel;
+
+        // std::printf("%f\r\n", rot_error);
+
+        const auto epsilon = 0.0001f;
+        // soccer tells us to "halt" by sending 0 vel commands, we want to
+        // freeze the
+        // rotational controller too so bots dont' react when getting handled
+        bool soccer_stop = (std::abs(target_vel_act[0]) < epsilon) &&
+                           (std::abs(target_vel_act[1]) < epsilon);
+        if (!soccer_stop && std::abs(target_vel_act[2]) < epsilon) {
+            if (!angle_hold) {
+                target_rotation = rotation;
+                angle_hold = true;
+            }
+
+            // get the smallest difference between two angles
+            float rot_error = std::atan2(std::sin(target_rotation - rotation),
```

Comment:
It might not matter. But the atan is probably a really computationally expensive way to do this especially since we don't have native floating point support. Probably. I think.

Maybe more ridiculous, but something like this might be faster.
double difference = secondAngle - firstAngle;
        while (difference < -180) difference += 360;
        while (difference > 180) difference -= 360;